### PR TITLE
batch_norm: SLM param cache for channels-last fwd 1D kernel, fix small-C perf

### PR DESCRIPTION
**Pull request by weishi-deng**
## Problem

The existing BatchNorm channels-last forward inference kernel uses a 2-D work-item layout
`(m, c/VEC)` where `m = N*H*W`.  For shapes where `C` is **not** a multiple of the
cache-line width in elements (32 for fp16 with VEC=2), consecutive work-items in the `m`
dimension share a cache line but write non-contiguous bytes, forcing the hardware to issue
a **read-modify-write round-trip** for every cache-line touched by the store.  The extra
DRAM traffic causes bandwidth efficiency (`eta = measured_BW / peak_BW`) to collapse well
below the hardware limit.

In addition, for shapes with small or non-power-of-2 channel counts the indirect gather
into the per-channel `(weight, bias, mean, var)` arrays creates irregular memory-access
patterns (low shader-fetch-ahead-hit rate, excess read-allocate traffic), compounding the
problem.

## Solution

1. **1-D linearised kernel** (`BatchNormTransformInputChannelsLast1DKernelFunctor`):
   Each work-item computes `flat_pos = global_id × VEC`.  32 consecutive lanes produce
   exactly 64 contiguous bytes — one full cache line — eliminating partial writes.

2. **SLM parameter cache** (`BatchNormTransformInputChannelsLast1DSLMKernelFunctor`):
   The four per-channel arrays (`weight`, `bias`, `mean`, `var`) are cooperatively
   loaded into SLM when they fit within the per-WG SLM budget:
   `4 * C * sizeof(accscalar_t) <= syclLocalMemSize() / concurrent_wgs`, where
   `concurrent_wgs = syclMaxWorkItemsPerTile() / syclDeviceMaxWorkGroupSize()`
   (40 on BMG).

3. **`IntDivider`**: div/mod to recover `(n, c, h, w)` from the flat index is replaced
   with fixed-point multiply-high arithmetic, reducing ALU pressure.

The old 2-D kernel functor is removed entirely; all channels-last forward transforms now
go through the unified 1-D path.

## Performance (Intel Arc B580, fp16, channels-last)

*`_native_batch_norm_legit_no_training`, 100 iterations after 20 warmup.*

*`T1 = 2 × sizeof(tensor) / 460 GB/s` — bandwidth-bound minimum time (B580 peak = 460 GB/s).*
*`eta = T1 / T2` where `T2` is the measured kernel time.*

*Before: PyTorch nightly `2.14.0.dev20260629+xpu`.
 After: this PR.*

### Large non-aligned C  (most impacted — partial write + large spatial dimensions)

| Shape | Before eta | After eta | Speedup |
|---|---|---|---|
| (1024, 228, 56, 56) | 0.145 | 0.863 | ×5.9 |
| (512,  316, 56, 56) | 0.161 | 0.860 | ×5.3 |
| (1024, 366, 28, 28) | 0.281 | 0.860 | ×3.1 |
| (2048, 200, 28, 28) | 0.339 | 0.859 | ×2.5 |
| (1024, 144, 56, 56) | 0.350 | 0.860 | ×2.5 |
| (1024, 240, 28, 28) | 0.419 | 0.861 | ×2.1 |
| (1024, 162, 56, 56) | 0.422 | 0.860 | ×2.0 |

### Small non-aligned even C

| Shape | Before eta | After eta | Speedup |
|---|---|---|---|
| (2048,  58, 28, 28) | 0.470 | 0.863 | ×1.8 |
| (1024,  36, 28, 28) | 0.548 | 0.840 | ×1.5 |
| (1024,  40, 28, 28) | 0.557 | 0.840 | ×1.5 |
| (1024,  72, 28, 28) | 0.567 | 0.854 | ×1.5 |
| (128,  108, 83, 83) | 0.564 | 0.865 | ×1.5 |

### Odd C

| Shape | Before eta | After eta | Speedup |
|---|---|---|---|
| (1024,  27, 56, 56) | 0.512 | 0.855 | ×1.7 |
| (1024,  61, 28, 28) | 0.535 | 0.843 | ×1.6 |
| (1024,  95, 14, 14) | 0.548 | 0.830 | ×1.5 |
| (1024, 117, 14, 14) | 0.555 | 0.836 | ×1.5 |

### Aligned C (C divisible by 32) — no regression

| Shape | Before eta | After eta |
|---|---|---|
| (1024, 128, 28, 28) | 0.841 | 0.857 |
| (2048, 256, 14, 14) | 0.852 | 0.857 |
| (1024, 512, 14, 14) | 0.860 | 0.865 |

### CODE REVIEW

File: src/ATen/native/xpu/sycl/BatchNormKernels.cpp [hunk 1]
```diff
@@ -1598,83 +1642,101 @@ void batch_norm_elemt_channels_last_template(
...
-                syclMaxWorkGroupSize(kfn),
-                reduction_size,
-                stride / VEC_SIZE,
-                false,
-                ELEMENTS_PER_WORK_ITEM);
-            auto global_range_vec = std::get<0>(config_vec);
-            auto local_range_vec = std::get<1>(config_vec);
-            sycl_kernel_submit(global_range_vec, local_range_vec, queue, kfn);
+          int64_t total_elems = (int64_t)reduction_size * stride;
+          if (VEC_SIZE == 2 && stride < 1024) {
```
Comment (jianyizh) [inline src/ATen/native/xpu/sycl/BatchNormKernels.cpp]: slm_per_wg = syclLocalMemSize() / concurrent_wgs

**Review by jianyizh — CHANGES_REQUESTED** [review]

File: src/ATen/native/xpu/sycl/BatchNormKernels.cpp [hunk 2]
```diff
@@ -1362,60 +1363,96 @@ void batch_norm_elemt_template(
...
+        accscalar_t m_c, inv_std_c, w_c, s_c;
+        load_params(c, m_c, inv_std_c, w_c, s_c);
+
+        auto tmp = w_c *
+                (static_cast<accscalar_t>(input_[idx]) - m_c) *
+                inv_std_c +
+            s_c;
+        if (z_ != nullptr)
+          tmp += z_[idx];
+        out_[idx] = (fuse_relu_ && tmp <= accscalar_t(0.0)
```
Comment (jianyizh) [inline src/ATen/native/xpu/sycl/BatchNormKernels.cpp]: tmp = tmp * (fuse_relu_ && tmp <= accscalar_t(0.0)) ? this can remove the if else branch, will it be faster

File: src/ATen/native/xpu/sycl/BatchNormKernels.cpp [hunk 3]
```diff
@@ -1362,60 +1363,96 @@ void batch_norm_elemt_template(
...
+      int total_elems = reduction_size_ * stride_;
+      int total_vecs  = total_elems / 2;
+
+      for (int id = item.get_global_id(0); id < total_vecs;
+           id += global_stride) {
+        int flat_pos = id * 2;  // always even → always 4-byte aligned
+
+        auto dm = vec_divider_.divmod(static_cast<unsigned int>(flat_pos));
+        int channel_0 = static_cast<int>(dm.mod);
+        int channel_1 = (channel_0 + 1 < stride_) ? channel_0 + 1 : 0;
```
Comment (jianyizh) [inline src/ATen/native/xpu/sycl/BatchNormKernels.cpp]: why not use for loop vec_size and unroll? maybe later someone want to add vec size 4?

File: src/ATen/native/xpu/sycl/BatchNormKernels.cpp [hunk 4]
```diff
@@ -1448,89 +1503,90 @@ struct BatchNormTransformInputChannelsLastKernelFunctor {
...
+    sycl::group_barrier(item.get_group());
 
-    int loop_count =
-        1 + (reduction_size_ - 1) / (inner_loop_stride * PARALLEL_LOADS);
-    int address_base = m_offset * stride_ + c_vec_offset;
-    int address_increment = inner_loop_stride * stride_;
+    // Phase 2: VEC=2 main loop (reads params from SLM).
+    int global_stride = static_cast<int>(item.get_global_range(0));
+    int total_elems   = reduction_size_ * stride_;
+    int total_vecs    = total_elems / 2;
```
Comment (jianyizh) [inline src/ATen/native/xpu/sycl/BatchNormKernels.cpp]: Here you always assume VEC=2. Do not hardcode 2. use constexpr VEC_SIZE=2

File: src/ATen/native/xpu/sycl/BatchNormKernels.cpp [hunk 5]
```diff
@@ -1448,89 +1503,90 @@ struct BatchNormTransformInputChannelsLastKernelFunctor {
...
-          using vec_t = memory::aligned_vector<scalar_t, VEC_SIZE>;
-          using layerscalar_vec_t =
-              memory::aligned_vector<layerscalar_t, VEC_SIZE>;
-          auto input_vec = *reinterpret_cast<vec_t*>(
-              const_cast<scalar_t*>(&input_[address_base]));
-          vec_t output_vec, z_vec;
-          layerscalar_vec_t weight_vec, shift_vec;
+    for (int id = static_cast<int>(item.get_global_id(0));
+         id < total_vecs; id += global_stride) {
+      int flat_pos = id * 2; // always even -> always 4-byte aligned
```
Comment (jianyizh) [inline src/ATen/native/xpu/sycl/BatchNormKernels.cpp]: hardcode here

File: src/ATen/native/xpu/sycl/BatchNormKernels.cpp [hunk 6]
```diff
@@ -1448,89 +1503,90 @@ struct BatchNormTransformInputChannelsLastKernelFunctor {
...
-            weight_vec = *reinterpret_cast<layerscalar_vec_t*>(
-                const_cast<layerscalar_t*>(&weight_[c_vec_offset]));
-          }
-          if (shift_ != nullptr) {
-            shift_vec = *reinterpret_cast<layerscalar_vec_t*>(
-                const_cast<layerscalar_t*>(&shift_[c_vec_offset]));
-          }
+      auto dm = vec_divider_.divmod(static_cast<unsigned int>(flat_pos));
+      int channel_0 = static_cast<int>(dm.mod);
+      int channel_1 = (channel_0 + 1 < stride_) ? channel_0 + 1 : 0;
```
Comment (jianyizh) [inline src/ATen/native/xpu/sycl/BatchNormKernels.cpp]: use for loop vec size and unroll?

File: src/ATen/native/xpu/sycl/BatchNormKernels.cpp [hunk 7]
```diff
@@ -1598,83 +1642,101 @@ void batch_norm_elemt_channels_last_template(
...
-  constexpr int VEC_SIZE = PREFERRED_VEC_SIZE;
+
+  int64_t wg_size =
+      std::min((int64_t)1024, syclDeviceMaxWorkGroupSize());
 
   if (input.scalar_type() != second_dtype) {
     AT_DISPATCH_FLOATING_TYPES_AND2(
         kHalf, kBFloat16, input.scalar_type(), "batchnorm_forward_xpu", [&] {
           using accscalar_t = at::acc_type_device<scalar_t, kXPU>;
+          constexpr int VEC_SIZE = sizeof(scalar_t) <= 2 ? 2 : 1;
```
Comment (jianyizh) [inline src/ATen/native/xpu/sycl/BatchNormKernels.cpp:1647]: check input ptr is 4 byte aligned here

**Review by jianyizh — CHANGES_REQUESTED** [review]
